### ISSUE TITLE
Improvements to involution DSL

### ISSUE THREAD

**Issue by marcoct** [probcomp/Gen.jl#230]
Some discussion in the PR for the involution DSL (https://github.com/probcomp/Gen/pull/216) and TODO notes from the original PR, spur suggestion of several possible improvements and extensions. I am putting them all in one issue now for conciseness -- if someone wants to address one of these, they can create a specific issue for it and link to/from this issue.

- Automating detection of discrete / continuous, and simplification of the syntax, resulting in 4 read/write macros, instead of the current 8. The copy macros already do not distinguish between the two.

- Adding the ability to check whether an address exists or not. (But without reading its value)

- Currently, the proposal_retval is not allowed to depend on the  values of any continuous random choices in the model or proposal trace, because otherwise continuous choice data leaks into the involution and avoids accounting by the Jacobian correction. Perhaps we could allow the dependency, while still ensuring the JAcobian is computed correctly, by using a call to `choice_gradients` on the proposal generative function?

- We could add more checks that we don't write to, or copy to, the same address twice (these can be only enabled in 'check' mode). Also, because it creates noninvertible Jacobian, it is invalid to copy from the same address twice and we could check for that to

- Users should be able to read and write multivariate random choices (i.e. array-valued) in the involution DSL. The implementation would probably need to flatten them alongside the scalar dimensions for input/output to f_array.

- Add the ability to pass two optional address namespaces to `@callinv`(one for the proposal, one for the model). Any addresses used by reads, writes or copies will occur under that address namespace.

- In the LaTex doc, explain relationship and differences from http://proceedings.mlr.press/v89/roberts19a.html

**Comment by bzinberg** [comment]
> Automating detection of discrete / continuous

I have strong reservations about this.  For example, a categorical whose values are specific floats seems like something that we do not want to forbid.  It could occur in e.g. a hand-implementation of a piecewise-uniform distribution.  I think this is similar to @marcoct's concern in https://github.com/probcomp/Gen/pull/216#issuecomment-608007854.

**Comment by alex-lew** [comment]
Yeah, I think ultimately the right way to handle that is to make more of this kind of information accessible in the trace -- though I think a feature like that would require careful design so that we don't compromise performance or place too high a burden on custom GFs.

**Comment by marcoct** [comment]
Relevant to this issue: I generalized the involution DSL to be useful in cases other than RJMCMC, including e.g. SMC with change-of-variables. In these settings, the function is not necessarily an involution, it is just a general bijection. For reference, this is happening on this branch: https://github.com/probcomp/Gen/tree/20200416-marcoct-translatordsl

This involves renaming the macro from `@involution` to `@bijection`, which is better because it better captures the essential feature that makes these functions easy to compute with. Also, it will be clear to others how this DSL relates to various notions of 'bijectors' in other systems (the answer is roughly, that it works with arbitrarily structured choice maps and can represent changes to the structure of choice maps, and that it has meaningful semantics for both discrete and continuous rvs, instead of just R^n to R^n functions).

~~@alex-lew Even though the non-top-level functions were previously not always an involution, I believe it may be the case (haven't checked this). that all callee functions in this case are bijections in some sense (e.g. partially applied).~~ Nevermind, that can't possibly be true, but maybe there's something there.

I haven't created a PR for this yet because I need to add some examples / tests and update the docs. This may not be done for a while.

**Comment by bzinberg** [comment]
Nice.  Btw, do you think it would be too pedantic to call these things diffeomorphisms instead of bijections?  For one data point, the term "bijection" does not invoke the notion of a change of variables for me -- when someone tells me that a function between R^d and R^d' is a "bijection," that actually makes me think it specifically _isn't_ smooth -- hence why the person didn't call it a diffeomorphism, or by a generic term like "mapping" (whereas "bijection" to me cues "this is only about sets, not about spaces with structure").

Also, are you familiar with `tfp.Bijector` (same comment on that name...), and if so, how does this compare?

**Comment by marcoct** [comment]
@bzinberg Re tfp.Bijector, yes I alluded to that and some other similar efforts like https://github.com/TuringLang/Bijectors.jl. As part of the documentation of this feature, I hope to do a comparison. I haven't looked closely at these others recently, but I anticipate that the main differences are that:

- it defines functions from key-value maps to key-value maps (i.e. choice maps) that represent execution traces of universal probabilistic programs, and not vector to vector transformations.
- it can to be used with a combination of discrete and real-valued inputs/outputs; note that the fact that it's a bijection plays a role here with respect to the discrete variables. In particular, you can write a bijection that only involves discrete variables, and the fact that it's a bijection in that setting is significant.
- it can be used to define changes in the structure of a trace (via the discrete variables)

Re diffeomorphism -- I agree that's more precise. But one reason I like 'bijection' is because it evokes tfp.Bijector and related things; which I believe are the most closely related constructs elsewhere, so it will be easier for users of these other tools to understand it's role. But I'm open to other names.

**Comment by bzinberg** [comment]
`DifferentiableBijection`?
I guess the shorter `SmoothBijection` might not make clear to everyone that we're talking about AD.

**Comment by bzinberg** [comment]
Maybe having the documentation say
> ### Defining Smooth Bijections Using the `@bijection` DSL
would be a good way to do it.  (And the class could be called `SmoothBijection`, and users would not have to touch that class directly.)
